File: Final_Project/Code/Project_4/Bai_toan_5/Exercises/9/CSS_bt5_ex9.py
class Tree:
    def __init__(self, n):
        # n: so nut trong cay
        # parent: mang cha
        # children: danh sach con
        self.n = n
        self.parent = [-1] * n
        self.children = [[] for _ in range(n)]

    def add_edge(self, u, v):
        # Them canh (u,v), tuc u la cha cua v
        self.parent[v] = u
        self.children[u].append(v)

    def previous_sibling(self, v):
        # Tim anh em truoc cua nut v
        if self.parent[v] == -1:
            return None
        siblings = self.children[self.parent[v]]
        idx = siblings.index(v)
        return siblings[idx - 1] if idx > 0 else None

# Vi du su dung
if __name__ == "__main__":
    t = Tree(5)
    t.add_edge(0, 1)
    t.add_edge(0, 2)
    t.add_edge(0, 3)
    print("Previous sibling of 2:", t.previous_sibling(2))  % 1
    print("Previous sibling of 1:", t.previous_sibling(1))  # None

File: Final_Project/Code/Project_4/Bai_toan_5/Exercises/9/CSS_bt5_ex9.cpp
#include <iostream>
#include <vector>
using namespace std;

class Tree {
public:
    int n; // so nut
    vector<int> parent; // mang cha
    vector<vector<int>> children; // danh sach con

    Tree(int vertices) : n(vertices) {
        parent.assign(n, -1);
        children.resize(n);
    }

    void add_edge(int u, int v) {
        // Them canh (u,v)
        parent[v] = u;
        children[u].push_back(v);
    }

    int previous_sibling(int v) {
        // Tim anh em truoc cua nut v
        if (parent[v] == -1) return -1;
        auto& siblings = children[parent[v]];
        for (int i = 0; i < siblings.size(); ++i) {
            if (siblings[i] == v) {
                return i > 0 ? siblings[i - 1] : -1;
            }
        }
        return -1;
    }
};

int main() {
    Tree t(5);
    t.add_edge(0, 1);
    t.add_edge(0, 2);
    t.add_edge(0, 3);
    cout << "Previous sibling of 2: " << t.previous_sibling(2) << endl; // 1
    cout << "Previous sibling of 1: " << t.previous_sibling(1) << endl; // -1
    return 0;
}